crates/test-utils/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+//! Test utilities shared between various crates.
+
+/// Shortcut for `<string>.parse().unwrap()`.
+#[macro_export]
+macro_rules! parse {
+    ($string:literal) => {
+        $string.parse().unwrap()
+    };
+}
+
+/// Macro for creating parametrized *synchronous* tests.
+///
+/// The `param_test!` macro accepts the name of an existing function, followed by a list of case
+/// names and their arguments. It expands to a module with a `#[test]` function for each of the
+/// cases. Each test case calls the existing, named function with their provided arguments.
+///
+/// See [`async_param_test`] for a similar macro that works with `async` function.
+///
+/// # Examples
+///
+/// Calling a simple test function can be done as follows
+///
+/// ```
+/// # use test_utils::param_test;
+/// #
+/// param_test! {
+///     test_sum: [
+///         positive_sums: (10, 7, 17),
+///         negative_sums: (-5, -3, -8)
+///     ]
+/// }
+/// fn test_sum(lhs: i32, rhs: i32, sum: i32) {
+///     assert_eq!(lhs + rhs, sum);
+/// }
+/// ```
+///
+/// Additionally, test functions can also have return types, such as a [`Result`]:
+///
+/// ```
+/// # use std::error::Error;
+/// # use test_utils::param_test;
+/// #
+/// param_test! {
+///     test_parses -> Result<(), Box<dyn Error>>: [
+///         positive: ("21", 21),
+///         negative: ("-17", -17)
+///     ]
+/// }
+/// fn test_parses(to_parse: &str, expected: i32) -> Result<(), Box<dyn Error>> {
+///     assert_eq!(expected, to_parse.parse()?);
+///     Ok(())
+/// }
+/// ```
+///
+/// Finally, attributes such as as `#[ignore]` may be added to individual tests:
+///
+/// ```
+/// # use std::error::Error;
+/// # use test_utils::param_test;
+/// #
+/// param_test! {
+///     test_parses -> Result<(), Box<dyn Error>>: [
```

Comment:
Suggestion: We define and use a `TestResult` type in several places. Maybe we can define it in this new crate?